Add gomod-sync

andrerfcsantos opened the pull request:
Part of #1881 

This PR introduces a new tool to the repo: `gomod-sync`. This is a little program to:

* List `go.mod` files and the Go versions they are declaring via the `gomod-sync list` command.
* Check if all `go.mod` files from the exercises are set to a specific version, with the `gomod-sync check` command. The desired version can be set with the `--goversion` flag.
* Update all `go.mod` files to a specific version with the `gomod-sync update` command. The desired version for the update can be set with the `--goversion` flag.

Right now these 3 commands are working. The `update` command was ran against the repo to update all exercises to 1.16, and the result is https://github.com/exercism/go/pull/1955

## Tasks

- [x] `list`, `check` and `update` commands
- [x] Setup the use of the `check` command in CI
- [x] Documentation for the tool
- [x] Add a way to allow specific go versions for specific exercises
- [x] Tests

andrerfcsantos commented:
Ready for review 🚀

This one is on the bigger side, so feel free to take your time, there is no rush.

This is what the output of the check looks like when it fails: https://github.com/exercism/go/runs/4662874598?check_suite_focus=true

For the output in case everything is fine, see the `go.mod check` below ⬇️

junedev commented on .github/workflows/gomod-sync.yml:
```diff
@@ -0,0 +1,27 @@
+  push:
+    branches:
+      - main
+      - master
```
Do we need this line?

andrerfcsantos replied:
Probably not. I copied that part from [exercises-tests.yml](https://github.com/exercism/go/blob/main/.github/workflows/exercise-tests.yml) where we also have that line.

junedev replied:
Probably from a time when there was a transition between main and master ...

andrerfcsantos replied:
Just removed and seems to be working fine.

junedev commented on .github/workflows/gomod-sync.yml:
```diff
@@ -0,0 +1,27 @@
+      - name: Check go.mod files
+        shell: bash
+        run: |
+          cd gomod-sync && go build
+          ./gomod-sync check --goversion 1.16 --exercises ../exercises
```
Couldn't we use `go run` here?

andrerfcsantos replied:
Yes, we can. I'll change it.

andrerfcsantos replied:
Done.

junedev commented on gomod-sync/README.md:
```diff
@@ -0,0 +1,101 @@
+# gomod-sync
+
+Utility tool to check and update the Go version specified in the `go.mod` files of all exercises.
+It works by specifying the desired Go version for all the `go.mod` files to be in: the `check` command
```
I don't understand the `:` after "to be in".

andrerfcsantos replied:
This seems like a copy-paste gone wrong. Thanks for pointing out. Will fix.

andrerfcsantos replied:
Turns out that was supposed to be a period `.`.

junedev commented:
Thanks a lot for taking care of this. I only read the README so far (which is great) and it sounds like it does all the great things we need it to do. However, I am contemplating whether it is possible to reduce the complexity of the tool a bit. You built something quite "general purpose" which is nice but since it also means more code, it means there is more code for future maintainers/contributors to read, understand and maintain. Also the various versions in which the command can be invoked make it a bit harder for users.

Here some concrete points:
* The tool could assume the default Go version is always provided via the config file, then you would not need the additional goversion flag. Currently, it is confusing what happens if the file says one thing and the flag says something else. So one way of doing things would reduce potential problems and complexity. The readme then would only say: "want to configure? change the config file".
* In our context, the folder that contains the exercises and the folder that contains the config will rarely ever change. It would be ok if those were hardcoded or part of the config file so they don't have to be passed when invoking the command. I think your tool is fast enough that we don't need the ability to one check individual exercise folders.
* Do we need the `list` command?

Point 1 and 2 together would then make the invocation as simple as `gomod-sync check` (same for the other commands) which would be a very user friendly API.

If you really want the tool to be used for other Go projects, then it would make more sense to keep in a separate repository and only `go get/install` it here.

junedev commented on gomod-sync/.gitignore:
```diff
@@ -0,0 +1,15 @@
+*.out
+
+# Dependency directories (remove the comment below to include it)
+# vendor/
```
Do we really need a separate gitignore file here? Couldn't we update the existing root level file instead?

andrerfcsantos replied:
I thought about it and I don't have a strong opinion. In the end I decided to put it here because I didn't want to pollute the global gitignore, and helps this folder to me more self-contained - we can move the folder around without the need to update the gitignore.

But I don't have a strong opinion either way.

andrerfcsantos replied:
I thought about it and I don't have a strong opinion. In the end I decided to put it here because I didn't want to pollute the global gitignore, and helps this folder to me more self-contained - we can move the folder around without the need to update the gitignore.

But I don't have a strong opinion either way.

junedev replied:
If you don't mind it would be great if you could remove this file and update the root gitignore instead. Imo, this is not polluting it but instead this file here mostly contains the standard entries that should be in the root level file anyway.

Also please include "gomod-sync" in the list of files to ignore so the binary that gets build does not show up in the files to commit.

andrerfcsantos replied:
Done.

junedev commented on gomod-sync/README.md:
```diff
@@ -0,0 +1,101 @@
+
+Available Commands:
+  check       Checks if all go.mod files are in the target version
+  completion  generate the autocompletion script for the specified shell
```
I don't understand what this does/means, can you help me? :thinking: :see_no_evil:

andrerfcsantos replied:
Which part? The `completion` command?

andrerfcsantos replied:
Which part? The `completion` command?

junedev replied:
Yes, the completion command.

andrerfcsantos replied:
Most shells like bash or zsh allow users to provide auto completion files that tell the shell how to autocomplete custom commands. The idea here is that if you write `gomod-sync ` and then press tab, your shell will autocomplete with some random file you have in the current directory, because the shell doesn't know better. What you want is the shell to cycle the list of commands that are specific to `gomod-sync` when you write `gomod-sync ` and then press tab. But for the shell to be able to do this, it has to know the syntax for `gomod-sync`, the list of commands,  flags, etc. These completion files allow a user to teach their shell about `gomod-sync` so it provides auto completions that make sense for the command.

What the `completion` command does is generate one of these completion files for `gomod-sync`. A user can then plug this file into their shell to provide auto completion for `gomod-sync`.

That command is implemented automatically by cobra. I think we can disable it if needed.

junedev replied:
Thanks for the info!

junedev commented on gomod-sync/README.md:
```diff
@@ -0,0 +1,101 @@
+
+* Check if all `go.mod` files of exercises in the `exercises` folder have the `1.16` Go version:
+
+  * `gomod-sync check --goversion 1.16 --exercises ./exercises`
```
If someone would just copy and paste this it would not work (because the executable does not exist). It would be good to give complete instructions that work for a new contributor that just cloned this repository. (Doesn't have to be here in "Examples", could also be some "install" part or similar.)

andrerfcsantos replied:
Sounds good!

andrerfcsantos replied:
Sounds good!

andrerfcsantos replied:
I added an install section. I also added a simpler example without any flags just to showcase the defaults values in action.

junedev commented on gomod-sync/.gitignore:
```diff
@@ -0,0 +1,15 @@
+*.out
+
+# Dependency directories (remove the comment below to include it)
+# vendor/
```
I think it would be good to include vendor in the gitignore just in case someone runs in vendored mode.

andrerfcsantos replied:
Sounds good!

andrerfcsantos replied:
Comment removed.

andrerfcsantos commented:
> Thanks a lot for taking care of this. I only read the README so far (which is great) and it sounds like it does all the great things we need it to do. However, I am contemplating whether it is possible to reduce the complexity of the tool a bit. You built something quite "general purpose" which is nice but since it also means more code, it means there is more code for future maintainers/contributors to read, understand and maintain. Also the various versions in which the command can be invoked make it a bit harder for users.

Thank you for reviewing! Specially such a big PR.

The tool ended being a bit bigger than I anticipated originally, but I think it's still at a manageable size. The alternative would be to have a bash script, but my experience is that bash scripts also become easily unmanageable when they grow a little. And with a language like Go we can at least write tests easily for it. So I believe it's a good trade-off. Plus, we know how to write Go, right? :P 

> Point 1 and 2 together would then make the invocation as simple as `gomod-sync check` (same for the other commands) which would be a very user friendly API.

You can already do this. If no flags are specified, it'll grab the target version from the config file. Not specifying the path to the exercises is also OK, as the default value for the flag is already hardcoded to be `..\exercises`. 

The reason I made the script use all the flags was just so it is more explicit about what the program is doing. The idea is that just by seeing the flags, you get a sense on what the program is trying to do, without the need to see the documentation or actually digging into the code to see how it works. But we can just remove the use of the flags without changing the program. I see merits in both approaches.

> * The tool could assume the default Go version is always provided via the config file, then you would not need the additional goversion flag. Currently, it is confusing what happens if the file says one thing and the flag says something else. So one way of doing things would reduce potential problems and complexity. The readme then would only say: "want to configure? change the config file".

If the version is not specified via the flag, it'll use the version in the configuration file.

If you specify the version flag, it will override the configuration file. I'm not sure if this is confusing, as it is the default behavior of every program I know that uses configuration files. I can make the documentation be more explicit about it though.

> * In our context, the folder that contains the exercises and the folder that contains the config will rarely ever change. It would be ok if those were hardcoded (for the config) or part of the config file (for the exercises) so they don't have to be passed when invoking the command. I think your tool is fast enough that we don't need the ability to one check individual exercise folders.

Both the path to the config file flag and the path to the exercises folder flag have [default values](https://github.com/andrerfcsantos/exercism-go/blob/3672d22c6a356b0110f21fdd5c4ddf690fc0f122/gomod-sync/cmd/root.go#L18) hardcoded, which means that you can omit both of those flags and the default value will be used.

The version flag has no default value, but the program always reads the config file to look for the target version, which means that the version flag can also be omitted.

> * Do we need the `list` command?

I used it in the beginning to get an idea of the global state of the repository, and I think it can be useful for that purpose. I don't think maintaining this command is expensive, as it makes use of a function that the `check` and `update` commands are also using and has only a for loop beyond that.

> If you really want the tool to be used for other Go projects, then it would make more sense to keep in a separate repository and only `go get/install` it here. However, that wouldn't solve the problem of keeping this as easy/maintainable as possible for Exercism folks.

Well, I don't want this tool to be used for other Go projects, so I think it's fine to leave it as is :) I don't think the tool is that general purpose in its current form, as it relies on having the folder structure we do and on the notion of exercises.

junedev commented:
> Thank you for reviewing! Specially such a big PR.

Don't thank me yet. :slightly_smiling_face: As I said, I have not reviewed the code, mainly only the README. (This is why I didn't know that you already had good defaults in place.)

I totally agree a bash script wouldn't be a good option. I was just asking myself whether we really need that much code for this task or whether there is significant overhead here because you wrote a proper command line application instead of a "go script" that does 2 things with some hard coded assumptions. The answer might very well be "no".

Let me try to group the other things we discussed above into some categories given the information you provided above. I am also ok if this gets merged without addressing any of the things mentioned below.

**I am strongly in favor of ...**
- Adding a section `## Quick Start` above `## Installing` section in the gomod-sync README that says something like this:
  ```
  To update all go.mod files according to the config file (gomod-sync/config.json) run `cd gomod-sync && go run main.go update`.
  
  To check all exercise go.mod files specify the correct Go version, run `cd gomod-sync && go run main.go check`.
  ```
- Add a section `## Managing the Go version` or similar to the root level readme (maybe below where the generator is explained) that contains the following
  - a very short explanation that we have the this tool, that it is built in Go and what it does
  - a link to the tool readme
  - the two lines mentioned above

**I am mid-level in favor of ...**
- Removing the list command entirely. Imo the less that needs to be explained in the readme and needs to be maintained in the code the better, even if its only some lines that are removed. In the future, the functionality of "getting an overview" is provided by looking at the config file.
- Removing the completion command from the readme. If there is really no extra code involved to support it, we could leave it as undocumented feature in case anyone ever tries it. Since the tool will be used by human like once a year and you need to type like 10 letters, I don't think anyone will miss the auto-completion thingy.

The two things above would make the command list `check`, `update`, `help` which would make it very clear what the tool is for.

**I am slightly in favor of but its totally up to you ...**
- Removing the flags

andrerfcsantos commented:
* Removed the `.gitignore` on the `gomod-sync` folder and moved the relevant settings to the global `.gitignore`
* Removed the `list` command
* Removed the `completion` command from the README
* Updated both README with the suggested changes

junedev commented on gomod-sync/gomod/version.go:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	// Write the file
+	_, err = file.WriteString(content)
```
What about https://pkg.go.dev/os#WriteFile instead of doing the file writing steps manually?

andrerfcsantos replied:
Done.

junedev reviewed (approved):
I tried to do a review. I don't know anything about cobra though so I can't comment on that. The part that I could understand looked good to me, just some minor comments.

andrerfcsantos commented on gomod-sync/cmd/config/load_test.go:
```diff
@@ -0,0 +1,82 @@
+			ExpectError: true,
+		},
+		{
+			Name: "Loading non-existent config",
```
It was indeed a copy/paste error. Well spotted, thanks! Fixed now.

andrerfcsantos commented on gomod-sync/cmd/config/version.go:
```diff
@@ -0,0 +1,60 @@
+
+// ExerciseVersion represents the version one particular exercise should
+// have in its go.mod file
+type ExerciseVersion struct {
```
Done.

andrerfcsantos commented on gomod-sync/cmd/config/version.go:
```diff
@@ -0,0 +1,60 @@
+// equalExceptions compares two lists of exercise versions and tells if they are equal.
+// Two exercise list versions are considered equal if they contain the same elements,
+// regardless of their order in both lists.
+func equalExceptions(a, b []ExerciseVersion) bool {
```
We don't strictly need it. At the time I thought it was a good idea to make this equal function not care about the order of the elements, but I understand it might complicate things unnecessarily. I removed that logic - the elements now must be in the same order in the JSON and in the config struct.

andrerfcsantos commented on gomod-sync/cmd/config/version_test.go:
```diff
@@ -0,0 +1,63 @@
+		Expected string
+	}{
+		{
+			Name: "Expected version just with a default should return the default",
```
I agree this was confusing. I changed this first case and reworded the others to make things more clear.

andrerfcsantos commented on gomod-sync/cmd/config/version.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// Equal tells if this VersionConfig is equal to another VersionConfig
+func (vc *VersionConfig) Equal(other VersionConfig) bool {
```
I had my doubts when writing this too.

Since we were indeed using these functions only in tests, I moved them to the relevant test files.

andrerfcsantos commented on gomod-sync/config.sample.json:
```diff
@@ -0,0 +1,9 @@
+{
```
I made this file inspired by other pieces of software I used before that provided a file like this. But this might not be necessary in our case since we have the README, and I understand how it can be confusing.

The file is now removed.

andrerfcsantos commented on gomod-sync/gomod/find.go:
```diff
@@ -0,0 +1,30 @@
+		return nil, fmt.Errorf("walking dir %q: %w", dirPath, err)
+	}
+
+	return files, err
```
:facepalm: Thanks :)

andrerfcsantos commented on gomod-sync/gomod/version.go:
```diff
@@ -0,0 +1,76 @@
+func init() {
+	// versionRegex is a regex that matches the Go version
+	// information in a go.mod file
+	versionRegex = regexp.MustCompile(`go (\d+\.\d+)`)
```
You are right. I modified it to check for the start and end of the line, optionally allowing for empty space at the beginning and  end of the line.

andrerfcsantos commented on gomod-sync/gomod/version.go:
```diff
@@ -0,0 +1,76 @@
+	defer file.Close()
+
+	// Read go.mod file
+	rawContent, err := io.ReadAll(file)
```
Done. There was one other place where this was present  - the `Update` function of this file. I fixed there too.

andrerfcsantos commented:
@junedev Thanks! Applied all the changes. Left comments in the conversations, feel free to un-resolve any of them or suggest further changes if there's something that can be further improved.

junedev commented:
@andrerfcsantos Thanks for addressing all the comments. Feel free to merge whenever you like. (I don't think there will be more reviews if we wait longer tbh.)
